pkg/exec/devfile.go

```diff
@@ -222,6 +224,61 @@ func ExecuteDevfileDebugActionWithoutRestart(client ExecClient, exec common.Exec
 	return nil
 }
 
+// ExecuteCompositeDevfileAction executes a given composite command in a devfile
+// The composite command may reference exec commands, composite commands, or both
+func ExecuteCompositeDevfileAction(client ExecClient, composite common.Composite, commandsMap map[string]common.DevfileCommand, compInfo adaptersCommon.ComponentInfo, show bool, machineEventLogger machineoutput.MachineEventLoggingClient) (err error) {
+	if composite.Parallel {
+		// Loop over each command and execute it in parallel
+		commandExecs := util.NewConcurrentTasks(len(composite.Commands))
+		for _, command := range composite.Commands {
+			cmd := command // needed to prevent the lambda from capturing the value
+			if devfileCommand, ok := commandsMap[strings.ToLower(cmd)]; ok {
+				commandExecs.Add(util.ConcurrentTask{ToRun: func(errChannel chan error) {
+					err := execCommandFromComposite(client, devfileCommand, commandsMap, compInfo, show, machineEventLogger)
```

Comment:
Just trying to think of all the edge cases, can we successfully exec in the same pod in parallel? Do we need mutex?

Reply:
Yeah, we can safely exec in the same pod in parallel.

Reply:
Granted, it may not be the best idea to have two commands exec'ing in parallel and accessing the same resource, but that's up to the devfile writer and is hard to guard against from our side.